compiler/src/main/scala/org/scalajs/nscplugin/GenJSExports.scala

```diff
@@ -697,57 +697,83 @@ trait GenJSExports[G <: Global with Singleton] extends SubComponent {
         previousArgsValues: Int => List[js.Tree])(
         implicit pos: Position): js.Tree = {
 
-      val trgSym = {
-        if (sym.isClassConstructor) {
+      val owner = sym.owner
+      val isNested = owner.isLifted && !isStaticModule(owner.originalOwner)
+
+      val (trgSym, trgTree) = {
+        if (!isNested && sym.isClassConstructor) {
+          assert(captures.isEmpty, "expected empty captures")
           /* Get the companion module class.
-           * For inner classes the sym.owner.companionModule can be broken,
-           * therefore companionModule is fetched at uncurryPhase.
+           * For classes nested inside of modules the sym.owner.companionModule
+           * can be broken, therefore companionModule is fetched at
+           * uncurryPhase.
+           */
+          val trgSym = enteringPhase(currentRun.namerPhase) {
+            owner.linkedClassOfClass
+          }
+          (trgSym, genLoadModule(trgSym))
+        } else if (!isNested && static) {
+          assert(captures.isEmpty, "expected empty captures")
+          (owner, genLoadModule(owner))
+        } else if (sym.isClassConstructor || static) {
+          assert(isNested, "unexpected isNested")
+          assert(captures.size == 1,
+              s"expected exactly one capture got $captures ($sym at $pos)")
+
+          /* Find the module accessor.
            *
            * #4465: If owner is a nested class, the linked class is *not* a
            * module value, but another class. In this case we need to call the
            * module accessor on the enclosing class to retrieve this.
+           *
+           * #4526: If the companion module is private, linkedClassOfClass
+           * does not work (because the module name is prefixed with the full
+           * path). So we find the module accessor first and take its result
+           * type to be the companion module type.
            */
-          val companionModule = enteringPhase(currentRun.namerPhase) {
-            sym.owner.companionModule
+
+          val outer = owner.originalOwner
+
+          val modAccessor = {
+            val name = enteringPhase(currentRun.typerPhase) {
+              owner.unexpandedName.toTermName
+            }
+
+            outer.info.members.find { sym =>
+              sym.isModule && sym.unexpandedName == name
+            }.getOrElse {
+              throw new AssertionError(
+                s"couldn't find module accessor for ${owner.fullName} at $pos")
+            }
+          }
+
+          val receiver = captures.head
+
+          val trgSym = modAccessor.tpe.resultType.typeSymbol
```

Comment:
IIUC, these are the only lines that have an actual change of behavior. Perhaps it would be worth it to decompose this commit in two:
1. The refactoring that computes `(trgSym, trgTree)` in one go, but preserving the behavior, and then
2. This change of behavior.

That might help further bisecting if we have to come back to this area.

Reply:
Hrm... If I do that, then I need to duplicate the logic of `trgSym` in the `if (isNested)` block (or introduce a change in behavior there already). Or am I missing something?

Reply:
Ah yes, you're right. Let's keep it as is, then.